src/common/snippets/src/lowered/pass/solve_buffer_memory.cpp

```diff
@@ -14,73 +15,144 @@ namespace snippets {
 namespace lowered {
 namespace pass {
 
-std::vector<ov::MemorySolver::Box> SolveBufferMemory::init_boxes(const AllocateBuffers::BufferClusters& buffer_clusters) {
-    std::vector<ov::MemorySolver::Box> boxes;
-    const auto count = static_cast<int>(buffer_clusters.size());
-    for (int i = 0; i < count; i++) {
-        ov::MemorySolver::Box box = { std::numeric_limits<int>::max(), 0, 0, i };
-        int64_t box_size = 0;
-        for (const auto& buffer_expr : buffer_clusters[i]) {
-            int e_start = 0, e_finish = 0;
-            const auto buffer = ov::as_type_ptr<ov::snippets::op::Buffer>(buffer_expr->get_node());
-            OPENVINO_ASSERT(buffer != nullptr, "BufferSolver expects Buffer ops in clusters");
-
-            // life finish time - order of LoopEnd / MemoryAccess ops
-            const auto& buffer_outs = buffer_expr->get_output_port_connectors();
-            for (const auto& buffer_out : buffer_outs) {
-                const auto consumers = buffer_out->get_consumers();
-                for (const auto& consumer : consumers) {
-                    const auto consumer_order = static_cast<int>(ov::snippets::pass::GetTopologicalOrder(consumer.get_expr()->get_node()));
-                    e_finish = std::max(e_finish, consumer_order);  // the last consumer
-                }
+std::pair<LinearIR::container, LinearIR::container> SolveBufferMemory::extract_static_and_dynamic_buffers(const LinearIR::container& buffer_expressions) {
+    LinearIR::container static_buffer_exprs, dynamic_buffer_exprs;
```

Comment:
Maybe store ptrs on `op::Buffer` instead of Expressions to avoid casting?

Reply:
We should have access to port connectors again [here](https://github.com/openvinotoolkit/openvino/blob/009c7fcf0fab42a4085bad4f7e5b2faec82be40a/src/common/snippets/src/lowered/pass/solve_buffer_memory.cpp#L57). To be honest, I  was thinking to create `BufferExpression` with the same interface of `Buffer` op to avoid many casting here 🤔